File: Spotify-ObjC/Spotify-ObjC/SARequestManager.m
//
//  SARequestManager.m
//  Spotify-ObjC
//
//  Created by Kevin Zeckser on 7/29/16.
//  Copyright © 2016 Kevin Zeckser. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "SARequestManager.h"

@interface SARequestManager ()

@end

@implementation SARequestManager

static NSURLSession *session;
static SARequestManager *sharedInstance = nil;

- (void) singletonInit {
    NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
    session = [NSURLSession sessionWithConfiguration: sessionConfig];
}

+ (SARequestManager *) sharedInstance {
    static dispatch_once_t dispatchOncePredicate = 0;
    dispatch_once(&dispatchOncePredicate, ^{
        sharedInstance = [[self alloc] init];
        [sharedInstance singletonInit];
    });
    return sharedInstance;
}

- (void) getItemsWithQuery:(NSString *)query completion:(Completion)completion {
    NSMutableString *path = [NSMutableString stringWithFormat: @"https://api.spotify.com/v1/search?q=%@", query];
    NSString *parameters = @"&type=track,artist&market=US";
    [path appendString: parameters];
    NSURL *url = [NSURL URLWithString: path];
    NSURLSessionDataTask *task = [session dataTaskWithURL: url
                                        completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                            SAResponse *result = [[SAResponse alloc] init];
                                            if (!data) {
                                                result.error = error;
                                                result.response = Failure;
                                            } else {
                                                result.response = Success;
                                                NSArray *itemsFromJSON = [SAArtist configureWithJSON:data];
                                                [result setItems: itemsFromJSON];
                                            }
                                            dispatch_sync(dispatch_get_main_queue(), ^{
                                                completion(result);
                                            });
                                        }];
    [task resume];
}

@end

File: Spotify-Swift/SpotifyArtistsViewer/SARequestManager.swift
//
//  SARequestManager.swift
//  SpotifyArtistsViewer
//
//  Created by Kevin Zeckser on 6/6/16.
//  Copyright © 2016 Kevin Zeckser. All rights reserved.
//
import Foundation

enum SAResponse {
    case Success(artists: [SAArtist])
    case Failure(error: ErrorType)
}

final class SARequestManager {
    
    // MARK: - Properties
    
    static let sharedInstance = SARequestManager()
    private let session: NSURLSession!
    private init() {
        let sessionConfig = NSURLSessionConfiguration.defaultSessionConfiguration()
        session = NSURLSession(configuration: sessionConfig)
    }
    
    // MARK: - NSURLRequest & JSON Parse
    
    func getArtistsWithQuery(query: String, completion: SAResponse -> Void) {
        let path = "https://api.spotify.com/v1/search?q=\(query)&type=track,artist&market=US"
        
        guard let url = NSURL(string: path) else { return }
        
        let task = session.dataTaskWithURL(url) { (data: NSData?, response: NSURLResponse?, error: NSError?) in
            var result: SAResponse
            if let error = error {
                result = SAResponse.Failure(error: error)
            } else if let data = data {
                do {
                    let jsonResult = try NSJSONSerialization.JSONObjectWithData(data, options: NSJSONReadingOptions.MutableContainers) as? NSDictionary
                    var returnedArtists = [SAArtist]()
                    if let artistResponse = jsonResult?["artists"],
                       let artists = artistResponse["items"] as? NSArray {
                        for entry in artists {
                            let artistEntry = entry as? NSDictionary
                            guard let name = artistEntry?["name"] as? String else { continue }
                            guard let images = artistEntry?["images"] as? NSArray else { continue }
                            guard let image = self.fetchImage(images as [AnyObject]) else { continue }
                            returnedArtists.append(SAArtist(name: name, image: image, description: nil))
                        }
                    }
                    result = SAResponse.Success(artists: returnedArtists)
                } catch let error as NSError {
                    result = SAResponse.Failure(error: error)
                }
            } else {
                result = SAResponse.Success(artists: [SAArtist]())
            }
            dispatch_async(dispatch_get_main_queue()) {
                completion(result)
            }
        }
        task.resume()
    }
    
    func fetchImage(images: [AnyObject]?) -> String?  {
        guard let images = images else { return nil }
        for image in images {
            guard ((image as? [String:AnyObject]) != nil) else { return nil }
            guard let imageDict = image as? [String:AnyObject] else { return nil }
            let height = imageDict["height"] as? Int
            let width = imageDict["width"] as? Int
            if height == width {
                if let url = image["url"] as? String {
                    return url
                }
            }
        }
        return nil
    }
}
